Crash in 3.27.0 when building Beam: "AnnotatedTypeMirrors aren't structurally equal."

I encountered a crash in checker framework 3.27.0

    git clone https://github.com/kennknowles/beam --branch checker-crash-FnApiDoFnRunner
    cd beam
    ./gradlew :sdks:java:harness:compileJava

```
error: AnnotatedTypeMirrors aren't structurally equal.
    type1 = AnnotatedTypeVariable( capture#657 extends @Initialized @NonNull BoundedWindow )
    type2 = AnnotatedDeclaredType( @Initialized @NonNull BoundedWindow )
    visitHistory = org.checkerframework.framework.type.StructuralEqualityVisitHistory@5d12c3d6
  ; The Checker Framework crashed.  Please report the crash.
  Compilation unit: /Users/klk/GitHub/apache/beam/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java
  Last visited tree at line 751 column 13:
              keyCoder,
  Exception: java.lang.Throwable; java.lang.Throwable
        at org.checkerframework.javacutil.BugInCF.<init>(BugInCF.java:22)
        at org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.defaultAction(AbstractAtmComboVisitor.java:52)
        at org.checkerframework.framework.type.StructuralEqualityComparer.defaultAction(StructuralEqualityComparer.java:56)
        at org.checkerframework.framework.type.StructuralEqualityComparer.defaultAction(StructuralEqualityComparer.java:29)
        at org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visitTypevar_Declared(AbstractAtmComboVisitor.java:543)
        at org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:587)
        at org.checkerframework.framework.type.StructuralEqualityComparer.areEqual(StructuralEqualityComparer.java:93)
        at org.checkerframework.framework.type.StructuralEqualityComparer.areEqualInHierarchy(StructuralEqualityComparer.java:105)
        at org.checkerframework.framework.type.DefaultTypeHierarchy.areEqualInHierarchy(DefaultTypeHierarchy.java:236)
        at org.checkerframework.framework.type.DefaultTypeHierarchy.isContainedBy(DefaultTypeHierarchy.java:311)
        at org.checkerframework.framework.type.DefaultTypeHierarchy.isContainedMany(DefaultTypeHierarchy.java:541)
        at org.checkerframework.framework.type.DefaultTypeHierarchy.visitTypeArgs(DefaultTypeHierarchy.java:518)
        at org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:448)
        at org.checkerframework.framework.type.DefaultTypeHierarchy.visitDeclared_Declared(DefaultTypeHierarchy.java:45)
        at org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:311)
        at org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:155)
        at org.checkerframework.framework.type.DefaultTypeHierarchy.isSubtype(DefaultTypeHierarchy.java:131)
        at org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:2787)
        at org.checkerframework.checker.nullness.NullnessVisitor.commonAssignmentCheck(NullnessVisitor.java:237)
        at org.checkerframework.common.basetype.BaseTypeVisitor.commonAssignmentCheck(BaseTypeVisitor.java:2764)
        at org.checkerframework.checker.nullness.NullnessVisitor.commonAssignmentCheck(NullnessVisitor.java:218)
        at org.checkerframework.common.basetype.BaseTypeVisitor.checkArguments(BaseTypeVisitor.java:3288)
        at org.checkerframework.common.basetype.BaseTypeVisitor.visitNewClass(BaseTypeVisitor.java:1933)
        at org.checkerframework.checker.nullness.NullnessVisitor.visitNewClass(NullnessVisitor.java:729)
        at org.checkerframework.checker.nullness.NullnessVisitor.visitNewClass(NullnessVisitor.java:68)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCNewClass.accept(JCTree.java:1727)
        at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
        at org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
        at org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:343)
        at org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:180)
        at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
        at jdk.compiler/com.sun.source.util.TreeScanner.visitAssignment(TreeScanner.java:585)
        at org.checkerframework.common.basetype.BaseTypeVisitor.visitAssignment(BaseTypeVisitor.java:1565)
        at org.checkerframework.common.basetype.BaseTypeVisitor.visitAssignment(BaseTypeVisitor.java:180)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCAssign.accept(JCTree.java:1904)
        at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
        at org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
        at org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:343)
        at org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:180)
        at jdk.compiler/com.sun.source.util.TreeScanner.visitExpressionStatement(TreeScanner.java:433)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCExpressionStatement.accept(JCTree.java:1476)
        at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
        at org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
        at org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:343)
        at org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:180)
        at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
        at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
        at jdk.compiler/com.sun.source.util.TreeScanner.visitBlock(TreeScanner.java:248)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCBlock.accept(JCTree.java:1048)
        at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
        at org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
        at org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:343)
        at org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:180)
        at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
        at jdk.compiler/com.sun.source.util.TreeScanner.visitMethod(TreeScanner.java:206)
        at org.checkerframework.framework.source.SourceVisitor.visitMethod(SourceVisitor.java:106)
        at org.checkerframework.common.basetype.BaseTypeVisitor.visitMethod(BaseTypeVisitor.java:949)
        at org.checkerframework.checker.initialization.InitializationVisitor.visitMethod(InitializationVisitor.java:334)
        at org.checkerframework.checker.nullness.NullnessVisitor.visitMethod(NullnessVisitor.java:511)
        at org.checkerframework.checker.nullness.NullnessVisitor.visitMethod(NullnessVisitor.java:68)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCMethodDecl.accept(JCTree.java:898)
        at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:82)
        at org.checkerframework.framework.source.SourceVisitor.scan(SourceVisitor.java:88)
        at org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:343)
        at org.checkerframework.common.basetype.BaseTypeVisitor.scan(BaseTypeVisitor.java:180)
        at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:90)
        at jdk.compiler/com.sun.source.util.TreeScanner.scan(TreeScanner.java:105)
        at jdk.compiler/com.sun.source.util.TreeScanner.scanAndReduce(TreeScanner.java:113)
        at jdk.compiler/com.sun.source.util.TreeScanner.visitClass(TreeScanner.java:187)
        at org.checkerframework.framework.source.SourceVisitor.visitClass(SourceVisitor.java:94)
        at org.checkerframework.common.basetype.BaseTypeVisitor.processClassTree(BaseTypeVisitor.java:533)
        at org.checkerframework.checker.initialization.InitializationVisitor.processClassTree(InitializationVisitor.java:288)
        at org.checkerframework.checker.nullness.NullnessVisitor.processClassTree(NullnessVisitor.java:575)
        at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:479)
        at org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:180)
        at jdk.compiler/com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:808)
        at jdk.compiler/com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:56)
        at org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:82)
        at org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:1010)
        at org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:542)
        at org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:188)
        at jdk.compiler/com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:828)
        at jdk.compiler/com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:132)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1414)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1361)
        at jdk.compiler/com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:960)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:104)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:147)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:100)
        at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:94)
        at org.gradle.internal.compiler.java.IncrementalCompileTask.call(IncrementalCompileTask.java:89)
        at org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.call(AnnotationProcessingCompileTask.java:94)
        at org.gradle.api.internal.tasks.compile.ResourceCleaningCompilationTask.call(ResourceCleaningCompilationTask.java:57)
        at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:54)
        at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:39)
        at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerWorkAction.execute(AbstractDaemonCompiler.java:135)
        at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:63)
        at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:49)
        at org.gradle.workers.internal.AbstractClassLoaderWorker$1.create(AbstractClassLoaderWorker.java:43)
        at org.gradle.internal.classloader.ClassLoaderUtils.executeInClassloader(ClassLoaderUtils.java:100)
        at org.gradle.workers.internal.AbstractClassLoaderWorker.executeInClassLoader(AbstractClassLoaderWorker.java:43)
        at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:32)
        at org.gradle.workers.internal.FlatClassLoaderWorker.run(FlatClassLoaderWorker.java:22)
        at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:87)
        at org.gradle.workers.internal.WorkerDaemonServer.run(WorkerDaemonServer.java:56)
        at org.gradle.process.internal.worker.request.WorkerAction$1.call(WorkerAction.java:138)
        at org.gradle.process.internal.worker.child.WorkerLogEventListener.withWorkerLoggingProtocol(WorkerLogEventListener.java:41)
        at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:135)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

So, this is actually a duplicate of https://github.com/typetools/checker-framework/issues/3785. ([Here's the relevant comment.](https://github.com/typetools/checker-framework/issues/3785#issuecomment-712519125)). However if you add a diamond operator to the constructor invocation:

```java
     this.stateAccessor =
         new FnApiStateAccessor<>(
             pipelineOptions,
             pTransformId,
             processBundleInstructionId,
             cacheTokens,
             bundleCache,
             processWideCache,
             tagToSideInputSpecMap,
             beamFnStateClient,
             keyCoder,
             windowCoder,
             this::getCurrentKey,
             () -> currentWindow);
```
Then Java isn't able to type check this code:
```
src/beam/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java:752: error: incompatible types: Coder<CAP#1> cannot be converted to Coder<BoundedWindow>
            windowCoder,
            ^
  where CAP#1 is a fresh type-variable:
    CAP#1 extends BoundedWindow from capture of ? extends BoundedWindow
```

So, while the Checker Framework should not crash on legal Java code, you can probably fix this crash by getting javac to compile this code using the diamond operator.

--- a/checker/tests/nullness/Issue3929.java
+++ b/checker/tests/nullness/Issue3929.java
@@ -12,7 +12,6 @@ public class Issue3929 {
 
   public void endElement(NullableMyClass3929 arg) {
     for (Object o : arg.getKeys()) {
-      // :: error: (dereference.of.nullable)
       o.toString();
     }
   }

--- a/checker/tests/nullness/MisuseProperties.java
+++ b/checker/tests/nullness/MisuseProperties.java
@@ -16,8 +16,7 @@ public class MisuseProperties {
     p.put("line.separator", null);
     Hashtable h = p;
     // Error, because HashTable value has NonNull bound.
-    //// :: error: (argument) :: warning: [unchecked] unchecked call to
-    //// put(K,V) as a member of the raw type java.util.Hashtable
+    // put(K,V) as a member of the raw type java.util.Hashtable
     // :: warning: [unchecked] unchecked call to put(K,V) as a member of the raw type
     // java.util.Hashtable
     // :: error: (argument)

--- a/checker/tests/nullness/MisuseProperties.java
+++ b/checker/tests/nullness/MisuseProperties.java
@@ -16,11 +16,11 @@ public class MisuseProperties {
     p.put("line.separator", null);
     Hashtable h = p;
     // Error, because HashTable value has NonNull bound.
-    // TODO: false negative. See #365.
     //// :: error: (argument) :: warning: [unchecked] unchecked call to
     //// put(K,V) as a member of the raw type java.util.Hashtable
     // :: warning: [unchecked] unchecked call to put(K,V) as a member of the raw type
     // java.util.Hashtable
+    // :: error: (argument)
     h.put("line.separator", null);
     // :: error: (argument)
     System.setProperty("line.separator", null);

--- a/checker/tests/nullness/RawTypesUses.java
+++ b/checker/tests/nullness/RawTypesUses.java
@@ -14,8 +14,7 @@ public abstract class RawTypesUses {
     @NonNull Object o1 = notRawNullable.foo();
 
     Generic rawNullable = new Generic<@Nullable String>();
-    // TODO: false negative. See #635.
-    //// :: error: (assignment)
+    // :: error: (assignment)
     @NonNull Object o2 = rawNullable.foo();
 
     Generic<@NonNull String> notRawNonNull = new Generic<>();
@@ -23,11 +22,9 @@ public abstract class RawTypesUses {
 
     Generic rawNonNull = new Generic<@NonNull String>();
     Generic rawNonNullAlais = rawNonNull;
-    // TODO: false negative. See #635.
-    //// :: error: (assignment)
+    // :: error: (assignment)
     @NonNull Object o4 = rawNonNull.foo();
-    // TODO: false negative. See #635.
-    //// :: error: (assignment)
+    // :: error: (assignment)
     @NonNull Object o5 = rawNonNullAlais.foo();
   }
 
@@ -40,18 +37,15 @@ public abstract class RawTypesUses {
     @NonNull Object o1 = notRawNullable.foo();
 
     Generic rawNullable = rawReturn();
-    // TODO: false negative. See #635.
-    //// :: error: (assignment)
+    // :: error: (assignment)
     @NonNull Object o2 = rawNullable.foo();
 
-    // TODO: false negative. See #635.
-    //// :: error: (assignment)
+    // :: error: (assignment)
     @NonNull Object o3 = rawReturn().foo();
 
     Generic local = rawReturn();
     Generic localAlias = local;
-    // TODO: false negative. See #635.
-    //// :: error: (assignment)
+    // :: error: (assignment)
     @NonNull Object o4 = local.foo();
   }
 }

--- a/checker/tests/nullness/generics/Issue269.java
+++ b/checker/tests/nullness/generics/Issue269.java
@@ -20,8 +20,7 @@ class Issue269 {
 
   void method2(CallbackNN callback) {
     // Forbid this call, because the bound is not respected.
-    // TODO: false negative. See #635.
-    //// :: error: (argument)
+    // :: error: (argument)
     // :: warning: [unchecked] unchecked call to handler(H) as a member of the raw type
     // Issue269.CallbackNN
     callback.handler(null);

--- a/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java
+++ b/framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java
@@ -3735,7 +3735,9 @@ public class BaseTypeVisitor<Factory extends GenericAnnotatedTypeFactory<?, ?, ?
             || memberReferenceTree.getTypeArguments().isEmpty())) {
       // Method type args
       requiresInference = true;
-    } else if (memberReferenceTree.getMode() == ReferenceMode.NEW) {
+    } else if (memberReferenceTree.getMode() == ReferenceMode.NEW
+        || ((JCTree.JCMemberReference) memberReferenceTree)
+            .hasKind(JCTree.JCMemberReference.ReferenceKind.UNBOUND)) {
       if (type.getKind() == TypeKind.DECLARED
           && ((AnnotatedDeclaredType) type).isUnderlyingTypeRaw()) {
         // Class type args

--- a/framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java
+++ b/framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java
@@ -440,6 +440,7 @@ public class AnnotatedTypes {
         }
         return result;
       case UNION:
+        return substituteTypeVariables(types, atypeFactory, receiverType, member, memberType);
       case DECLARED:
         AnnotatedDeclaredType receiverTypeDT = (AnnotatedDeclaredType) receiverType;
         if (receiverTypeDT.isUnderlyingTypeRaw()

--- a/framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java
+++ b/framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java
@@ -441,7 +441,19 @@ public class AnnotatedTypes {
         return result;
       case UNION:
       case DECLARED:
-        if (((AnnotatedDeclaredType) receiverType).isUnderlyingTypeRaw()) {
+        AnnotatedDeclaredType receiverTypeDT = (AnnotatedDeclaredType) receiverType;
+        if (receiverTypeDT.isUnderlyingTypeRaw()
+            && member
+                .getEnclosingElement()
+                .equals(receiverTypeDT.getUnderlyingType().asElement())) {
+          // Section 4.8, "Raw Types".
+          // (https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.8)
+          //
+          // The type of a constructor (§8.8), instance method (8.4, 9.4), or non-static field
+          // (8.3) of a raw type C that is not inherited from its superclasses or superinterfaces
+          // is the raw type that corresponds to the erasure of its type in the generic declaration
+          // corresponding to C.
+
           return memberType.getErased();
         }
         return substituteTypeVariables(types, atypeFactory, receiverType, member, memberType);

--- a/framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java
+++ b/framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java
@@ -441,6 +441,9 @@ public class AnnotatedTypes {
         return result;
       case UNION:
       case DECLARED:
+        if (((AnnotatedDeclaredType) receiverType).isUnderlyingTypeRaw()) {
+          return memberType.getErased();
+        }
         return substituteTypeVariables(types, atypeFactory, receiverType, member, memberType);
       default:
         throw new BugInCF("asMemberOf called on unexpected type.%nt: %s", receiverType);
@@ -696,6 +699,10 @@ public class AnnotatedTypes {
     // Has the user supplied type arguments?
     if (!targs.isEmpty()) {
       List<? extends AnnotatedTypeVariable> tvars = preType.getTypeVariables();
+      if (tvars.isEmpty()) {
+        // This happens when the method is invoked with a raw receiver.
+        return Collections.emptyMap();
+      }
 
       Map<TypeVariable, AnnotatedTypeMirror> typeArguments = new HashMap<>();
       for (int i = 0; i < elt.getTypeParameters().size(); ++i) {

--- a/framework/tests/all-systems/Issue3785.java
+++ b/framework/tests/all-systems/Issue3785.java
@@ -1,0 +1,39 @@
+import java.util.List;
+
+public class Issue3785<T> {
+  public Issue3785(List<T> l) {}
+
+  <L> void method(L l) {}
+
+  @SuppressWarnings("unchecked")
+  void use(Issue3785 p, Object o, List<String> list) {
+    // This type-checks in Java, but probably shouldn't.
+    p.<Void>method(o);
+    p.method(o);
+    new Issue3785(list);
+  }
+
+  /*
+  <L> L method2(L l) { return l;}
+
+  @SuppressWarnings("unchecked")
+  void use2(Issue3785<String> p, Object o) {
+      // javac issues: "error: incompatible types: Object cannot be converted to Void"
+      p.<Void>method(o);
+      // javac issues: "error: incompatible types: Object cannot be converted to Void"
+      Void s = p.<Void>method2(o);
+  }
+  */
+}
+
+/*
+class Issue3785NonParameterized {
+    <L> void method(L l) {}
+
+    @SuppressWarnings("unchecked")
+    void use(Issue3785NonParameterized p, Object o) {
+        // javac issues: "error: incompatible types: Object cannot be converted to Void"
+        p.<Void>method(o);
+    }
+}
+*/

--- a/framework/tests/all-systems/Issue5436.java
+++ b/framework/tests/all-systems/Issue5436.java
@@ -1,0 +1,26 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+@SuppressWarnings("unchecked")
+public class Issue5436 {
+  static class BoundedWindow {}
+
+  static class Accessor<K> {
+    public Accessor(Supplier<BoundedWindow> s) {}
+
+    Accessor() {}
+  }
+
+  private final Accessor<?> stateAccessor;
+  private List<BoundedWindow> currentWindows;
+
+  Issue5436() {
+    currentWindows = new ArrayList<>();
+    stateAccessor = new Accessor(() -> currentWindows);
+  }
+
+  private Object getCurrentKey() {
+    return new Object();
+  }
+}

--- a/framework/tests/framework/BridgeMethods.java
+++ b/framework/tests/framework/BridgeMethods.java
@@ -20,8 +20,7 @@ class Usage {
     c.id(new @Odd Object());
 
     // Oddness is wrong! Would also fail with ClassCastException.
-    // TODO: false negative. See #635.
-    //// :: error: (argument)
+    // :: error: (argument)
     // :: warning: [unchecked] unchecked call to id(T) as a member of the raw type C
     // :: warning: (cast.unsafe.constructor.invocation)
     c.id(new @Even Object());

